### PR TITLE
Implementation of `GET /throttles` for WebService

### PR DESCRIPTION
**Pull request by garyparrot**
這個 PR 幫 web service 實作取得 throttle 設定的功能

```java
  @Test
  void testThrottleHandler1() {
    try (Admin admin = Admin.of(bootstrapServers())) {
      var handler = new ThrottleHandler(admin);
      var topicName = Utils.randomString();
      admin.creator().topic(topicName).numberOfPartitions(3).numberOfReplicas((short) 3).create();
      Utils.sleep(Duration.ofSeconds(1));
      admin
          .replicationThrottler()
          .throttle(topicName)
          .ingress(DataRate.MiB.of(500).perSecond())
          .egress(DataRate.MiB.of(500).perSecond())
          .apply();
      Utils.sleep(Duration.ofSeconds(1));

      System.out.println(handler.get(Optional.empty(), Map.of()).json());
    }
  }
```

<details>
{
  "brokers": {
    "0": {
      "ingress": 524288000,
      "egress": 524288000
    },
    "1": {
      "ingress": 524288000,
      "egress": 524288000
    },
    "2": {
      "ingress": 524288000,
      "egress": 524288000
    }
  },
  "topics": [
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 1,
      "broker": 1,
      "type": "leader"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 0,
      "broker": 1,
      "type": "follower"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 1,
      "broker": 2,
      "type": "follower"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 0,
      "broker": 2,
      "type": "leader"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 0,
      "broker": 0,
      "type": "follower"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 2,
      "broker": 2,
      "type": "follower"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 2,
      "broker": 0,
      "type": "leader"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 1,
      "broker": 0,
      "type": "follower"
    },
    {
      "name": "a401a89996e0449eada1f85d2b2dc128",
      "partition": 2,
      "broker": 1,
      "type": "follower"
    }
  ]
}

<details>

### CODE REVIEW

File: app/src/test/java/org/astraea/app/web/ThrottleHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          .replicationThrottler()
+          .throttle(topicName)
+          .ingress(DataRate.MiB.of(500).perSecond())
+          .egress(DataRate.MiB.of(500).perSecond())
+          .apply();
+      Utils.sleep(Duration.ofSeconds(1));
+
+      System.out.println(handler.get(Optional.empty(), Map.of()).json());
+    }
+  }
```
Comment (garyparrot) [inline app/src/test/java/org/astraea/app/web/ThrottleHandlerTest.java:48]: 晚點會補上比較詳細的測試

File: app/src/main/java/org/astraea/app/web/ThrottleHandler.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    private ThrottleSetting(
+        Map<Integer, Map<ThrottleBandwidths, Long>> brokers, Collection<ThrottleTarget> topics) {
+      this.brokers = brokers;
+      this.topics = topics;
+    }
+
+    @Override
+    public String json() {
+      return new GsonBuilder()
+          .registerTypeAdapter(Optional.class, new SerializeOptional())
```
Comment (chia7712) [inline app/src/main/java/org/astraea/app/web/ThrottleHandler.java:168]: 可否把這個實作移到上層去，讓所有物件都可以享有這個好處

File: app/src/main/java/org/astraea/app/web/ThrottleHandler.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            .collect(Collectors.toUnmodifiableSet());
+
+    return new ThrottleSetting(bandwidths, simplify(leaderTargets, followerTargets));
+  }
+
+  /** break apart the {@code throttled.replica} string setting into a set of throttle targets */
+  private Set<TopicPartitionReplica> toReplicaSet(String topic, String throttledReplicas) {
+    if (throttledReplicas.isEmpty()) return Set.of();
+
+    if (throttledReplicas.equals("*"))
```
Comment (chia7712) [inline app/src/main/java/org/astraea/app/web/ThrottleHandler.java:102]: 這個還是留個TODO好了，當作後續增強

**Review by chia7712 — COMMENTED** [review]
@garyparrot 感謝這隻patch，概念很棒，幾個小建議請看一下

**Comment by chia7712** [comment]
@garyparrot 抱歉，麻煩修正一下衝突

File: app/src/main/java/org/astraea/app/web/ThrottleHandler.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    return new ThrottleSetting(bandwidths, simplify(leaderTargets, followerTargets));
+  }
+
+  /**
+   * break apart the {@code throttled.replica} string setting into a set of topic/partition/replicas
+   */
+  private Set<TopicPartitionReplica> toReplicaSet(String topic, String throttledReplicas) {
+    if (throttledReplicas.isEmpty()) return Set.of();
+
+    // TODO: support for wildcard throttle might be implemented in the future.
```
Comment (chia7712) [inline app/src/main/java/org/astraea/app/web/ThrottleHandler.java:99]: 可否先把議題開好？並且在這裡給予連結

File: app/src/main/java/org/astraea/app/web/Response.java [hunk 5]
```diff
@@ -50,7 +55,11 @@ default int code() {
   }
 
   default String json() {
-    return new Gson().toJson(this);
+    return new GsonBuilder()
```
Comment (chia7712) [inline app/src/main/java/org/astraea/app/web/Response.java:58]: 這裡有一些延伸的議題：

1. `RecordHandler`也有自定義，是否可以整合？
2. java Optional 系列的類別是否都要整合？

麻煩開一個議題，並且留個TODO

File: app/src/main/java/org/astraea/app/web/ThrottleHandler.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  }
+
+  static class ThrottleTarget {
+    final String name;
+    final OptionalInt partition;
+    final OptionalInt broker;
+    final Optional<LogIdentity> type;
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (chia7712) [inline app/src/main/java/org/astraea/app/web/ThrottleHandler.java:171]: 這是為了測試才寫的嗎？

**Review by chia7712 — COMMENTED** [review]
@garyparrot 大致完美，幾個小建議請看一下

File: app/src/main/java/org/astraea/app/web/ThrottleHandler.java (hunk 6 above)
Comment (garyparrot) [reply app/src/main/java/org/astraea/app/web/ThrottleHandler.java:171]: 這是因為 `ThrottleTarget` 會被放在 `Set` 裡面所以才實作 `equals` 和 `hashCode`

測試的部分，是因為 IntelliJ 的 test coverage tool 說那裡面有幾行測試每有跑到，我想說順手補上讓整個檔案變綠的，所以才添加的。

**Pull request merged**